dali/test/python/test_operator_cast.py

```diff
@@ -12,46 +12,106 @@
-        self.layout = "HWC"
-        self.device = device
-        self.iterator = iterator
-        self.inputs = ops.ExternalSource()
-        self.cast = [ops.Cast(device=device, dtype=dtype) for dtype in cast_dtypes]
-
-    def define_graph(self):
-        self.data = self.inputs()
-        out = self.data.gpu() if self.device == 'gpu' else self.data
-        for k in range(len(self.cast)):
-            out = self.cast[k](out)
-        return out
-
-    def iter_setup(self):
-        data = self.iterator.next()
-        self.feed_input(self.data, data, layout=self.layout)
-
-def check_cast_operator_float16(device, batch_size, in_type, out_type):
-    input_shape=(300, 400, 3)
-    eii1 = RandomlyShapedDataIterator(batch_size, max_shape=input_shape, dtype=in_type)
-    eii2 = RandomlyShapedDataIterator(batch_size, max_shape=input_shape, dtype=in_type)
-    compare_pipelines(
-        CastPipeline(device, batch_size, iter(eii1), [types.FLOAT16, out_type]),
-        CastPipeline(device, batch_size, iter(eii2), [out_type]),
-        batch_size=batch_size, N_iterations=5)
-
-def test_cast_operator_float16():
+from test_utils import check_batch, np_type_to_dali
+
+
+def ref_cast(x, dtype):
+    if np.issubdtype(dtype, np.integer):
+        lo = np.iinfo(dtype).min
+        hi = np.iinfo(dtype).max
+        if np.issubdtype(x.dtype, np.floating):
+            x = np.round(x)
+        else:
+            x
+        return x.clip(lo, hi).astype(dtype)
+    else:
+        return x.astype(dtype)
+
+def random_shape(rng, ndim:int, max_size:int):
+    if ndim == 0:
+        return []
+    max_size = int(max_size ** (1/ndim))
+    return list(rng.integers(0, max_size, [ndim]))
+
+def generate(rng, ndim:int, batch_size:int, in_dtype:np.dtype, out_dtype:np.dtype):
+    lo, hi = -1000, 1000
+    if np.issubdtype(out_dtype, np.integer):
+        lo = np.iinfo(out_dtype).min
+        hi = np.iinfo(out_dtype).max
+        if hi < np.iinfo(np.int64).max:
+            r = hi - lo
+            hi += r // 2
+            lo -= r // 2
+        if np.issubdtype(in_dtype, np.integer):
+            lo = max(np.iinfo(in_dtype).min, lo)
+            hi = min(np.iinfo(in_dtype).max, hi)
+        else:
+            lo = max(-np.finfo(in_dtype).max, lo)
+            hi = min(np.finfo(in_dtype).max, hi)
+
+    max_size = 100000 // batch_size
+    out = [rng.uniform(lo, hi, size=random_shape(rng, ndim, max_size)).astype(in_dtype) for _ in range(batch_size)]
+    if np.issubdtype(in_dtype, np.floating) and np.issubdtype(out_dtype, np.integer):
+        for x in out:
+            # avoid exactly halfway numbers - rounding is different for CPU and GPU
+            halfway = x[x - np.floor(x) == 0.5]
+            x[x - np.floor(x) == 0.5] = np.nextafter(halfway, np.Infinity)
+    return out
+
+rng = np.random.default_rng(1234)
+
+@nottest
+def _test_operator_cast(ndim, batch_size, in_dtype, out_dtype, device):
+    src = lambda: generate(rng, ndim, batch_size, in_dtype, out_dtype)
```

Comment:
I need to pass a _function_, not the results.

Reply:
How about making it more pythonic, by making a `generate` actually a generator and passing a generator object to `external_source`?
```python
def generate(...):
   yield out

fn.external_source(generate(...))
```

Reply:
Ultimately, the problem lies not with whether this is a generator or a callback, but that I have to pass parameters to it. I could probably use `partial` instead of a lambda - although this is also just syntactic sugar and it's a bit of a stretch to use `partial` for _full_ parameterization (but I'm not sure whether ES would handle it properly).

Reply:
> an you have an infinite generator?

Do you mean something like that?
```
def gen():
  while True:
    yield sample
```